### PR TITLE
Add WorkflowContext methods to Workflow CRD

### PR DESCRIPTION
**Pull request by micahhausler**
Signed-off-by: Micah Hausler <mhausler@amazon.com>

## Description

* Add WorkflowContext methods to Workflow type
* Add WorkflowContext conversion func
* Relocate FrozenTime for usage outside `/pkg`

## Why is this needed

This creates reusable logic for the gRPC server and other K8s client code (like boots).

Fixes: #

## How Has This Been Tested?

Added unit tests for new methods and functions

## How are existing users impacted? What migration steps/scripts do we need?

N/A

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [x] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: pkg/apis/core/v1alpha1/workflow_methods.go [hunk 1]
```diff
@@ -30,3 +30,83 @@ func (w *Workflow) GetStartTime() *metav1.Time {
...
+				action.Status == "STATE_FAILED" || action.Status == "STATE_TIMEOUT") && !found {
+				taskIndex = ti
+				actionTaskIndex = ai
+				found = true
+			}
+			if !found {
+				actionIndex++
+			}
+		}
+	}
```
Comment (detiber) [inline pkg/apis/core/v1alpha1/workflow_methods.go:67]: I was having some difficulty reading through the original logic, so I took a swag at trying to make it a bit more readable, not sure if I succeeded, though.

```suggestion
	taskIndex, actionIndex, actionTaskIndex, actionCount := func (tasks []Tasks) (int, int, int, int) {
		var (
			found           bool
			taskIndex       = -1
			actionIndex     int
			actionTaskIndex int
			actionCount     int
		)
		
		for ti, task := range w.Status.Tasks {
			actionCount += len(task.Actions)
			
			if found {
				continue
			}
			
			for ai, action := range task.Actions {
				switch action.Status {
				case "STATE_SUCCESS":
					actionIndex++
					continue	
				case "STATE_PENDING", "STATE_RUNNING", "STATE_FAILED", "STATE_TIMEOUT":
					taskIndex = ti
					actionTaskIndex = ai
					found = true
					break
				}
			}
		}
		
		return taskIndex, actionIndex, actionTaskIndex, actionCount
	}()
```

File: pkg/apis/core/v1alpha1/workflow_methods.go [hunk 2]
```diff
@@ -30,3 +30,83 @@ func (w *Workflow) GetStartTime() *metav1.Time {
...
+		actionIndex     int
+		actionTaskIndex int
+		actionCount     int
+	)
+	for ti, task := range w.Status.Tasks {
+		for ai, action := range task.Actions {
+			actionCount++
+			// Find either the first action in a non-terminal state or the first action in an unsuccessful terminal state
+			if (action.Status == "STATE_PENDING" || action.Status == "STATE_RUNNING" ||
+				action.Status == "STATE_FAILED" || action.Status == "STATE_TIMEOUT") && !found {
```
Comment (detiber) [inline pkg/apis/core/v1alpha1/workflow_methods.go:58]: There should likely be constants for these statuses here.

Comment (micahhausler) [reply pkg/apis/core/v1alpha1/workflow_methods.go:58]: Sure, I can add these

File: pkg/apis/core/v1alpha1/workflow_methods.go (hunk 1 above)
Comment (micahhausler) [reply pkg/apis/core/v1alpha1/workflow_methods.go:67]: Good suggestion, and the tests pass with this. thank you!

**Comment by mmlb** [comment]
I love the stuff in workflow_methods.go. I've been growing increasingly more annoyed at slinging pb types around instead of proper go types and this file makes it more obvious why. Going to create an issue for dropping that and maybe lifting some of the stuff from workflow_methods.go over and out of just here.

**Review by mmlb — APPROVED** [review]

**Review by mmlb — APPROVED** [review]

**Pull request merged**